### PR TITLE
Fix auc calculation and add tests

### LINKED ISSUE #195: AUC computation is incorrect for tensors with >2048 elements.
## 🐛 Bug

When `auc` is called with `reorder=True`, a stable sort is performed on the input tensors using `_stable_1d_sort` ([code](https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/functional/classification/auc.py#L38)). However, `_stable_1d_sort` returns at most `2048` elements if `nb` is not provided ([code](https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/utilities/data.py#L179)). This leads to `auc` computation being incorrect.

### PR DESCRIPTION
**Pull request by maximsch2**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes #195
Closes #196

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by maximsch2** [comment]
Stable sort that we are trying to do doesn't work: https://colab.research.google.com/gist/maximsch2/4341c3b01482f8b38ea0f0ed70affd33/stable-sort-in-pytorch.ipynb

File: torchmetrics/utilities/data.py [hunk 1]
```diff
@@ -151,35 +151,6 @@ def get_num_classes(
     return num_classes
```
Comment (ananyahjha93) [inline torchmetrics/utilities/data.py:153]: @SkafteNicki @Borda tagging you both before making the decision to completely remove ``_stable_1d_sort``. Can't see the usage anywhere else.

**Review by ananyahjha93 — APPROVED** [review]

File: tests/classification/test_auc.py [hunk 2]
```diff
@@ -32,19 +32,29 @@ def sk_auc(x, y):
     return _sk_auc(x, y)
 
 
+def sk_auc_reorder(x, y):
+    x = x.flatten()
+    y = y.flatten()
+    idx = np.argsort(x, kind='stable')
+    x = x[idx]
+    y = y[idx]
+    return _sk_auc(x, y)
```
Comment (SkafteNicki) [inline tests/classification/test_auc.py:41]: lets combine these two into one function with an `reorder=True/False` arg
```python
def sk_auc(x,y,reorder=False):
    x = x.flatten()
    y = y.flatten()
    if reorder:
        idx = np.argsort(x, kind='stable')
        x = x[idx]
        y = y[idx]
    return _sk_auc(x,y)

File: tests/classification/test_auc.py [hunk 3]
```diff
@@ -62,8 +72,14 @@ def test_auc(self, x, y, ddp, dist_sync_on_step):
...
-    def test_auc_functional(self, x, y):
-        self.run_functional_metric_test(x, y, metric_functional=auc, sk_metric=sk_auc, metric_args={"reorder": False})
+    @pytest.mark.parametrize("reorder", [True, False])
+    def test_auc_functional(self, x, y, reorder):
+        if reorder:
+            self.run_functional_metric_test(x, y, metric_functional=auc,
+                                            sk_metric=sk_auc_reorder, metric_args={"reorder": reorder})
+        else:
+            self.run_functional_metric_test(x, y, metric_functional=auc, sk_metric=sk_auc,
+                                            metric_args={"reorder": reorder})
```
Comment (SkafteNicki) [inline tests/classification/test_auc.py:82]:
```suggestion
        self.run_functional_metric_test(x, y, metric_functional=auc,
                                        sk_metric=partial(sk_auc_reorder, reorder=reorder), 
                                        metric_args={"reorder": reorder})
```

File: torchmetrics/functional/classification/auc.py [hunk 4]
```diff
@@ -35,7 +33,7 @@ def _auc_update(x: Tensor, y: Tensor) -> Tuple[Tensor, Tensor]:
 
 def _auc_compute(x: Tensor, y: Tensor, reorder: bool = False) -> Tensor:
     if reorder:
-        x, x_idx = _stable_1d_sort(x)
+        x, x_idx = torch.sort(x)
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/auc.py:36]: `torch.sort` is not stable, meaning that you will get different result if evaluating on cpu vs gpu:
```python
import torch
x = torch.randint(5, (10,))
torch.sort(x)
#torch.return_types.sort(
#values=tensor([0, 0, 1, 1, 1, 2, 3, 4, 4, 4]),
#indices=tensor([0, 2, 3, 4, 7, 9, 8, 1, 5, 6]))
torch.sort(x.cuda())
#torch.return_types.sort(
#values=tensor([0, 0, 1, 1, 1, 2, 3, 4, 4, 4], device='cuda:0'),
#indices=tensor([0, 2, 7, 4, 3, 9, 8, 1, 5, 6], device='cuda:0'))
```
this is the reason that we implemented a stable sorting algorithm

**Comment by SkafteNicki** [comment]
@maximsch2 I see your point that no matter what we do, we will never get stable sorting on cpu until torch v1.9 is released. Could you please add a note to the affected metrics, that users can expect slightly different results on cpu vs gpu. 
Also add a `TODO` that this is something we probably should revisit in the future.

File: tests/classification/test_auc.py [hunk 5]
```diff
@@ -32,19 +32,29 @@ def sk_auc(x, y):
...
-for i in range(4):
-    x = np.random.randint(0, 5, (NUM_BATCHES * 8))
-    y = np.random.randint(0, 5, (NUM_BATCHES * 8))
-    idx = np.argsort(x, kind='stable')
-    x = x[idx] if i % 2 == 0 else x[idx[::-1]]
-    y = y[idx] if i % 2 == 0 else x[idx[::-1]]
-    x = x.reshape(NUM_BATCHES, 8)
-    y = y.reshape(NUM_BATCHES, 8)
-    _examples.append(Input(x=tensor(x), y=tensor(y)))
+for batch_size in (8, 4049):
```
Comment (Borda) [inline tests/classification/test_auc.py:48]: that will be very long can we iterate in some steps - 10 or 100?

File: tests/classification/test_auc.py (hunk 2 above)
Comment (Borda) [inline tests/classification/test_auc.py:41]:
```suggestion
def sk_auc(x, y, reorder=False):
    x = x.flatten()
    y = y.flatten()
    if reorder:
        idx = np.argsort(x, kind='stable')
        x = x[idx]
        y = y[idx]
    return _sk_auc(x, y)
```
follow https://github.com/PyTorchLightning/metrics/pull/197/files#r619033509

File: tests/classification/test_auc.py [hunk 6]
```diff
@@ -62,8 +72,11 @@ def test_auc(self, x, y, ddp, dist_sync_on_step):
             dist_sync_on_step=dist_sync_on_step,
         )
 
-    def test_auc_functional(self, x, y):
-        self.run_functional_metric_test(x, y, metric_functional=auc, sk_metric=sk_auc, metric_args={"reorder": False})
+    @pytest.mark.parametrize("reorder", [True, False])
+    def test_auc_functional(self, x, y, reorder):
+        self.run_functional_metric_test(x, y, metric_functional=auc,
+                                        sk_metric=partial(sk_auc_reorder, reorder=reorder),
```
Comment (Borda) [inline tests/classification/test_auc.py:78]:
```suggestion
                                        sk_metric=partial(sk_auc, reorder=reorder), 
```

File: torchmetrics/functional/classification/auc.py (hunk 4 above)
Comment (SkafteNicki) [inline torchmetrics/functional/classification/auc.py:36]:
```suggestion
        # TODO: include stable=True arg when pytorch v1.9 is released
        x, x_idx = torch.sort(x)
```

File: torchmetrics/utilities/data.py (hunk 1 above)
Comment (SkafteNicki) [reply torchmetrics/utilities/data.py:153]: fine by me :]

File: tests/classification/test_auc.py (hunk 5 above)
Comment (SkafteNicki) [reply tests/classification/test_auc.py:44]: @Borda this is fine, it is not `range(8, 4049)` but just iteration over the two values.

**Review by SkafteNicki — APPROVED** [review]

**Pull request merged**